src/nvim/ex_cmds.c

```diff
@@ -4157,6 +4156,17 @@ static buf_T *do_sub(exarg_T *eap, proftime_T timeout,
             nmatch = vim_regexec_multi(&regmatch, curwin, curbuf,
                                        sub_firstlnum, matchcol, NULL, NULL);
 
+          if (!preview || has_second_delim) {
+            if (!did_save) {
```

Comment:
instead of convoluting the code like this you could just add `curbuf->deleted_bytes2 = 0` here